pkg/profiles/tap.go

```diff
@@ -0,0 +1,136 @@
+package profiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// RenderTapOutputProfile performs a tap on the desired resource and generates
+// a service profile with routes pre-populated from the tap data
+// Only inbound tap traffic is considered.
+func RenderTapOutputProfile(client pb.ApiClient, tapResource, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int, w io.Writer) error {
+	requestParams := util.TapRequestParams{
+		Resource:  tapResource,
+		Namespace: namespace,
+	}
+	log.Debugf("Running `linkerd tap %s --namespace %s`", tapResource, namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+
+	profile, err := tapToServiceProfile(client, req, namespace, name, controlPlaneNamespace, tapDuration, routeLimit)
+	if err != nil {
+		return err
+	}
+
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int) (sp.ServiceProfile, error) {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace),
+			Namespace: controlPlaneNamespace,
```

Comment:
FYI this will semantically conflict with https://github.com/linkerd/linkerd2/pull/2200.  Depending on which change lands first, someone will need to resolve this to point at the service's namespace.